resolwe_bio/tools/expression_fpkm_tpm.R

```diff
@@ -12,31 +13,36 @@ args = parser$parse_args(commandArgs(trailingOnly=TRUE))
 
 # retrieve exon lengths
 annot <- import.gff(args$annotation, format=args$format, genome="NA", feature.type=args$feature_type)
-exon_lengths=width(annot)
-names(exon_lengths)=elementMetadata(annot)[,args$id_attribute]
+exon_lengths <- width(annot)
+names(exon_lengths) <- elementMetadata(annot)[, args$id_attribute]
 
-# calculate gene lengths
-exon_lengths_by_gene=split(exon_lengths, names(exon_lengths))
-gene_lengths=sapply(exon_lengths_by_gene, sum)
+# calculate gene lengths, note that the result is lexicographically sorted
+gene_lengths <- sapply(split(exon_lengths, names(exon_lengths)), FUN = sum)
 
 # normalise counts by library size
-counts <- read.delim(args$counts_file, header = FALSE)
-rownames(counts) <- counts[,1]
-counts[,1] <- NULL
-lib_size=colSums(counts)
-ncounts=t(t(counts)/lib_size)
+ncounts <- read.delim(args$counts_file, header = FALSE)
+names(ncounts) <- c("Gene", "raw_count")
+ncounts$norm_count <- ncounts$raw_count / sum(ncounts$raw_count) # library_size
 
 # obtain CPM
-cpm=ncounts*1e6
+cpm <- ncounts$norm_count * 1e6
+names(cpm) <- ncounts$Gene
+cpm <- cpm[order(names(cpm))] # lexicographically order cpm
 
 # normalise counts by gene length
```

Comment:
In the next two lines after this comment you don't normalise by gene length, but you are selecting genes in counts table that you will normalise.

Reply:
Thanks, changed.